kroxylicious-wasm-sample-poc/src/main/java/io/kroxylicious/sample/util/SampleWasmFilterTransformer.java

```diff
@@ -0,0 +1,131 @@
+ * transformation of data in ProduceRequests and FetchResponses.
+ */
+public class SampleWasmFilterTransformer {
+
+    /**
+     * Transforms the given partition data according to the provided configuration.
+     * @param partitionData the partition data to be transformed
+     * @param context the context
+     * @param config the transform configuration
+     */
+    public static void transform(ProduceRequestData.PartitionProduceData partitionData, FilterContext context, SampleFilterConfig config) {
+        partitionData.setRecords(transformPartitionRecords((AbstractRecords) partitionData.records(), context, config.getReplacerModule()));
+    }
+
+    /**
+     * Transforms the given partition data according to the provided configuration.
+     * @param partitionData the partition data to be transformed
+     * @param context the context
+     * @param config the transform configuration
+     */
+    public static void transform(FetchResponseData.PartitionData partitionData, FilterContext context, SampleFilterConfig config) {
+        partitionData.setRecords(transformPartitionRecords((AbstractRecords) partitionData.records(), context, config.getReplacerModule()));
+    }
+
+    /**
+     * Performs find-and-replace transformations on the given partition records.
+     * @param records the partition records to be transformed
+     * @param context the context
+     * @param replacerModule the wasm replacer module to be used
+     * @return the transformed partition records
+     */
+    private static AbstractRecords transformPartitionRecords(AbstractRecords records, FilterContext context, String replacerModule) {
+        if (records.batchIterator().hasNext()) {
+            ByteBufferOutputStream stream = context.createByteBufferOutputStream(records.sizeInBytes());
+            MemoryRecordsBuilder newRecords = createMemoryRecordsBuilder(stream, records.firstBatch());
+
+            for (RecordBatch batch : records.batches()) {
+                for (Record batchRecord : batch) {
+                    newRecords.append(batchRecord.timestamp(), batchRecord.key(), transformRecord(batchRecord.value(), replacerModule),
+                            batchRecord.headers());
+                }
+            }
+            return newRecords.build();
+        }
+        return records;
+    }
+
+    // Hacked around caching of last used Wasm Module
+    private static String currentWasmModule;
+    private static Instance currentWasmInstance;
+    private static ExportFunction replacerFunction;
+    private static ExportFunction alloc;
+    private static ExportFunction dealloc;
+    private static Memory memory;
+
+    /**
+     * Performs a find-and-replace transformation of a given record value.
+     * @param in the record value to be transformed
+     * @param replacerModule the value to be replaced
+     * @return the transformed record value
+     */
+    private static ByteBuffer transformRecord(ByteBuffer in, String replacerModule) {
```

Comment:
This is the core of the change, here we load the WASM module(if it was not the last one used), and we run it with the provided input.
At the moment the scope is pretty limited, but it demonstrates the feasibility of the solution.